/* Copyright (c) 2020 mission systems pty ltd */

syntax = "proto2";
import "goby/protobuf/option_extensions.proto";
import "dccl/option_extensions.proto";
import "goby/acomms/protobuf/driver_base.proto"; // load up message DriverBaseConfig
import "goby/acomms/protobuf/modem_message.proto";

package goby.acomms.popoto.protobuf;

message Config
{
    message ModemAddress
    {
        required string ip = 2 [default = "127.0.0.1"];
        required string port = 3 [default = "17000"];
    }

    optional int32 start_timeout = 1 [
        default = 30,
        (goby.field).description =
            "Number of seconds to initialize the modem before giving up"
    ];

    optional int32 payload_mode = 2 [
        default = 0,
        (dccl.field).min = 0,
        (dccl.field).max = 5,
        (goby.field).description =
            "The modulation rate of the payload portion of the waveform, 0-5"
    ];

    optional int32 modem_power = 3 [
        default = 1,
        (dccl.field).min = 1,
        (dccl.field).max = 100,
        (goby.field).description = "Modem Tx strength 1-100W"
    ];

    required ModemAddress local = 4 [(goby.field).description =
                                     "The IP and port of the local modem "
                                     "(default port is 17000)"];
}
// added 1442 in modem driver
extend goby.acomms.protobuf.DriverConfig
{
    optional Config config = 1442;
}

enum TransmissionType
{
    BASE_TYPE = 0;
    POPOTO_TWO_WAY_PING = 1;  // modem 1 interrogates modem 2; modem 2 replies
                              // and modem 1 computes one way travel time
    POPOTO_PLAY_FILE = 2;
    POPOTO_TWO_WAY_RANGE = 3;
    POPOTO_DEEP_SLEEP = 4;
    POPOTO_WAKE = 5;
}
message RangingReply
{
    required double one_way_travel_time = 1
        [(dccl.field).units.unit = "si::second"];
}
message Transmission
{
    optional TransmissionType type = 1 [
        (goby.field).description =
            "Type of transmission if base `type` == DRIVER_SPECIFIC",
        default = BASE_TYPE
    ];
    optional RangingReply ranging_reply = 3;
    optional string file_location = 4;
    optional double transmit_power = 5;
}

extend goby.acomms.protobuf.ModemTransmission
{
    optional Transmission transmission = 1441;
}
